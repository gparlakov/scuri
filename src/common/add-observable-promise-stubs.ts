import { EOL } from 'os';
import { ConstructorParam, DependencyMethodReturnTypes } from '../types';

export function addDefaultObservableAndPromiseToSpyJoined(p: ConstructorParam, deps?: DependencyMethodReturnTypes, options?: DefaultMethodReturnsOptions): string {
    if(!deps?.has(p.type)) {
        return ''
    }
    const joiner = typeof options?.joiner === 'string' ? options?.joiner : EOL;
    
    return `${typeof joiner === 'string'? joiner : ''}${addDefaultObservableAndPromiseToSpy(p, deps, options).join(joiner)}`;
}

export function addDefaultObservableAndPromiseToSpy(p: ConstructorParam, deps?: DependencyMethodReturnTypes, options?: DefaultMethodReturnsOptions): string[] {
    if(!deps?.has(p.type)) {
        return []
    }
    const spyReturn = (v: string) => options?.spyReturnType === 'jest' ? `.mockReturnValue(${v})` : `.and.returnValue(${v})`;
    const dep = deps.get(p.type);
    const observables = Array.from(dep!.entries())
        .filter(([_, value]) => value.match(/Observable<|Subject</))
        .map(([key,]) => `${p.name}.${key}${spyReturn('EMPTY')}`);

    const promises = Array.from(dep!.entries())
        .map(([key, value]) => {
            return [key, value];
        })
        .filter(([_, value]) => value.match(/Promise</))
        .map(([key]) => `${p.name}.${key}${spyReturn('new Promise(res => {})')}`);
    return [...observables, ...promises];
}

interface DefaultMethodReturnsOptions {
    joiner?: string;
    spyReturnType?: 'jasmine' | 'jest';
}

export function listAllDefaultReturns(params: ConstructorParam[], deps?: DependencyMethodReturnTypes, options?: DefaultMethodReturnsOptions) {
    return params.flatMap(p => addDefaultObservableAndPromiseToSpy(p, deps, options))
}