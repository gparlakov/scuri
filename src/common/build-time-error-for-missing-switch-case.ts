export function buildErrorForMissingSwitchCase(_: never, description: string): never {
    throw new Error(description);
}
