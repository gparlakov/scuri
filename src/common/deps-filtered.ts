import {
    DependencyCallDescription,
    DependencyMethodReturnAndPropertyTypes,
} from '../types';

export function depsFiltered(
    deps: DependencyMethodReturnAndPropertyTypes | undefined,
    skip: (d: DependencyCallDescription) => boolean
): DependencyMethodReturnAndPropertyTypes | undefined {
    if (deps == null || !(deps instanceof Map)) {
        return deps;
    }
    // prettier-ignore
    return new Map(
        [...deps.entries()]
            .map(([key, value]) => ([
                key, new Map(
                    [...value.entries()].filter(([, dep]) => !skip(dep))
                )
            ])
        )
    );
}


export function dependenciesWrap(ds: DependencyMethodReturnAndPropertyTypes | undefined) {

    return {
        skip(skipWhere: (d: DependencyCallDescription) => boolean) {
            return depsFiltered(ds, skipWhere)
        }
    }
}
