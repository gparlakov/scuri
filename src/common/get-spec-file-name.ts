import { basename, join, normalize } from '@angular-devkit/core';
import { applyPathTemplate, FileEntry, TEMPLATE_FILENAME_RE } from '@angular-devkit/schematics';
import { ClassTemplateData } from '../types';

export function getSpecFilePathName(name: string) {
    return name.replace('.ts', '.spec.ts');
    // const normalizedName = normalize(name);
    // const ext = extname(basename(normalizedName));

    // return name.split(ext)[0] + '.spec' + ext;
}

export function getSpecFileCustomName(
    classData: Omit<ClassTemplateData, 'levenshtein' | 'params' | 'publicMethods' | 'setupMethods'>,
    templateName: string
): string | undefined {
    const templateFile: FileEntry = {
        path: basename(normalize(templateName)),
        content: Buffer.from(''),
    };

    const fileResult = applyPathTemplate(classData)(templateFile);

    const folderfyFn = folderfy(classData.folder);

    return typeof fileResult?.path === 'string'
        ? folderfyFn(
              fileResult.path
                  // remove template if any
                  .replace(TEMPLATE_FILENAME_RE, '')
          )
        : undefined;
}

function folderfy(folder: string) {
    return (fileName: string) => {
        return join(
            normalize(folder),
            fileName
                // remove the path as a possible duplicate
                .replace(folder, '')
        );
    };
}
