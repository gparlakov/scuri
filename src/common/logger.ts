import { Logger } from '@angular-devkit/core/src/logger';

let mainLogger: Logger
export function setLogger(l: Logger) {
    if(mainLogger != null) {
        // logger already set - don't overwrite
        return;
    }
    mainLogger = l;
}

export function getLogger(name: string): Logger {
    if(mainLogger == null) {
        setLogger(new Logger('just a logger'));
    }
    return mainLogger.createChild(name);
}
