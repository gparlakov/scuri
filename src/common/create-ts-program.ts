import { Tree } from '@angular-devkit/schematics';
import * as ts from 'typescript';
import { getLogger } from './logger';

let programCache = new Map<string, ts.Program>();
export function createTsProgram(fileName: string, tree: Tree): ts.Program {
    const logger = getLogger(createTsProgram.name);
    logger.debug(`entering for ${fileName}`)

    if(programCache.has(fileName)) {
        logger.debug(`Reusing program for ${fileName}`)
        return programCache.get(fileName)!;
    }

    let config: ts.ParsedCommandLine = {
        options: {
            target: ts.ScriptTarget.ES2015,
            module: ts.ModuleKind.CommonJS
        },
        fileNames: [],
        errors: [],
    };

    const closestTsConfig = ts.findConfigFile(fileName, (f) => tree.exists(f));
    if (typeof closestTsConfig === 'string') {
        logger.debug(`Found ${closestTsConfig}`);
        const src = ts.readJsonConfigFile(closestTsConfig, (f) => {
            if (tree.exists(f)) {
                const raw = tree.read(f);
                if (raw != null) {
                    return raw.toString();
                }
            }
        });
        logger.debug(`Parsed json ${src?.getText() ?? 'TsConfigSourceFile is nullish'}`);
        config = ts.parseJsonSourceFileConfigFileContent(
            src,
            {
                fileExists: (f) => {
                    logger.debug(`parseConfig: fileExists ${f}?${tree.exists(f)}`);
                    return tree.exists(f);
                },
                useCaseSensitiveFileNames: true,
                readFile: (f) => {
                    logger.debug(`parseConfig: readFile ${f} : ${tree.read(f)?.toString()}`);
                    return tree.read(f)?.toString();
                },
                readDirectory: (r, _exs, _excludes, _includes, _depth) => {
                    logger.debug(`parseConfig:readDirectory ${r}`);
                    return tree.getDir(r).subfiles.map((s) => s);
                },
            },
            tree.root.path,
            undefined,
            closestTsConfig
        );

        logger.debug(`parsed config: ${JSON.stringify(config)}`);
    } else {
        let checkedFiles: string[] = [];
        ts.findConfigFile(fileName, (f) => (checkedFiles.push(f), tree.exists(f)));
        logger.warn(
            `Did not find a tsconfig file close to ${fileName}. Some types might be missing. Searched in ${checkedFiles.join(
                '||'
            )}`
        );
    }

    const defaultHost = ts.createCompilerHost({}, true);
    logger.debug(
        `Creating (and caching) program for ${fileName}, with options\n${JSON.stringify(config.options)}`
    );
    const program = ts.createProgram([fileName], config.options, defaultHost);
    logger.debug(`FileContents: ${program.getSourceFile(fileName)?.getFullText()}`)
    programCache.set(fileName, program)
    return program;
}

export function getSourceFile(fullFileName: string, tree: Tree): ts.SourceFile | undefined {
    const logger = getLogger(getSourceFile.name);
    logger.debug(`entering for ${fullFileName}`)
    // const program = createTsProgram(fullFileName, tree);

    // if(program == null) {
    //     logger.debug('Program/source file is still null, probably some error in log above. Creating a new source file from scratch.')
    //     const contents = tree.read(fullFileName)?.toString();
    //     if(contents == null) {
    //         logger.error(`The contents of the ${fullFileName} are empty - the file is probably missing. Can't create a source file`)
    //         return undefined;
    //     }
    //     logger.debug(`Creating just a source file for ${fullFileName}`)
    return ts.createSourceFile(fullFileName, tree.read(fullFileName)?.toString() ?? '', ts.ScriptTarget.Latest, true);
    // }

    // logger.debug('Found program. Returning the source file from it')
    // return program.getSourceFile(fullFileName);
}
// failed experiment - faking the ts host by reading/writing from the tree host
// failed b/c tree does not get node_modules and other  necessary .d.ts files that comprise a big chunk of the base types like string, Number, HTMLElement, etc.
// keeping it around as it took way too much effort to just delete... :(

// export function createTsProgram_1(fileName: string): ts.Program {
//     const logger = getLogger(createTsProgram.name);

//     let config: ts.ParsedCommandLine = {
//         options: {
//             target: ts.ScriptTarget.ES2015,
//             module: ts.ModuleKind.CommonJS,
//         },
//         fileNames: [],
//         errors: [],
//     };

//     const closestTsConfig = ts.findConfigFile(fileName, (f) => tree.exists(f));
//     if (typeof closestTsConfig === 'string') {
//         logger.debug(`Found ${closestTsConfig}`);
//         const src = ts.readJsonConfigFile(closestTsConfig, (f) => {
//             if (tree.exists(f)) {
//                 const raw = tree.read(f);
//                 if (raw != null) {
//                     return raw.toString();
//                 }
//             }
//         });
//         logger.debug(`Parsed json ${src?.getText() ?? 'TsConfigSourceFile is nullish'}`);
//         config = ts.parseJsonSourceFileConfigFileContent(
//             src,
//             {
//                 fileExists: (f) => {
//                     logger.debug(`parseConfig: fileExists ${f}?${tree.exists(f)}`);
//                     return tree.exists(f);
//                 },
//                 useCaseSensitiveFileNames: true,
//                 readFile: (f) => {
//                     logger.debug(`parseConfig: readFile ${f} : ${tree.read(f)?.toString()}`);
//                     return tree.read(f)?.toString();
//                 },
//                 readDirectory: (r, _exs, _excludes, _includes, _depth) => {
//                     logger.debug(`parseConfig:readDirectory ${r}`);
//                     return tree.getDir(r).subfiles.map((s) => s);
//                 },
//             },
//             tree.root.path,
//             undefined,
//             closestTsConfig
//         );

//         logger.debug(`parsed config: ${JSON.stringify(config)}`);
//     } else {
//         let checkedFiles: string[] = [];
//         ts.findConfigFile(fileName, (f) => (checkedFiles.push(f), tree.exists(f)));
//         logger.warn(
//             `Did not find a tsconfig file close to ${fileName}. Some types might be missing. Searched in ${checkedFiles.join(
//                 '\n'
//             )}`
//         );
//     }

//     const defaultHost = ts.createCompilerHost({}, true);
//     const treeHost = <ts.CompilerHost>{
//         ...defaultHost,
//         readFile: (s: string) => {
//             const res = tree.read(s)?.toString();
//             logger.debug(`treeHost: readFile ${s}: ${res}`);
//             return res;
//         },
//         writeFile: (name: string, text: string) => {
//             logger.debug(`treeHost: writing ${name}:${text}`);
//             tree.overwrite(name, text);
//         },
//         fileExists: (name: string) => {
//             logger.debug(`treeHost: exists ${name}:${tree.exists(name)}`);
//             return tree.exists(name);
//         },
//         getNewLine: () => EOL,
//         getCurrentDirectory: () => {
//             logger.debug(`treeHost: getCurrentDirectory ${tree.root.path}}`);
//             return tree.root.path
//         },
//         getDirectories: (path: string) => {
//             logger.debug(`treeHost: getDirectories for ${path}: ${tree.getDir(path)?.subdirs}`);
//             return tree.getDir(path)?.subdirs
//         },
//         directoryExists: (directoryName: string) => {
//             logger.debug(`treeHost: directoryExists for ${directoryName}: ${tree.getDir(directoryName) != null}`);
//             return tree.getDir(directoryName) != null
//         },
//         readDirectory(rootDir, extensions, excludes, includes, depth?) {
//             logger.debug(`treeHost: readDirectory for, ${rootDir}, ${extensions}, ${excludes}, ${includes}, ${depth}`);
//             return tree.getDir(rootDir).subfiles
//         },
//     };
//     logger.debug(
//         `Creating program for ${fileName}, with options\n${JSON.stringify(config.options)}`
//     );
//     return ts.createProgram([fileName], config.options, treeHost);
// }


