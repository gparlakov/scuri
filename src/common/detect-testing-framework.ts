import { Tree } from '@angular-devkit/schematics';
import { getLogger } from './logger';

export type Supported = 'jest' | 'jasmine';
const supported: Supported[] = ['jest', 'jasmine'];

/**
 * Take the framework from the 1. cli params or 2. config or 3. best effort autodetect
 * @param cliParam command line interface param if any
 * @param configParam configured param if any
 * @param t Tree
 * @param fallback - if all else fails use this
 * @returns 'jest' | 'jasmine'
 */
export function detectTestingFramework(
    cliParam: string | undefined,
    configParam: string | undefined,
    t: Tree,
    fallback: Supported
): 'jest' | 'jasmine' {
    const logger = getLogger(detectTestingFramework.name);

    logger.log('debug', 'entering');
    if (typeof cliParam === 'string' && supported.includes(cliParam as Supported)) {
        logger.debug(`Detected supported cli param ${cliParam}`);
        return cliParam as Supported;
    }
    if (typeof cliParam === 'string') {
        logger.warn(
            `Detected un-supported cli param ${cliParam}. please use ${supported.join(' or ')}`
        );
    }
    if (typeof configParam === 'string' && supported.includes(configParam as Supported)) {
        logger.debug(`Detected supported config param ${cliParam}`);
        return configParam as Supported;
    }
    if (typeof configParam === 'string') {
        logger.warn(
            `Detected un-supported config param ${configParam}. please use ${supported.join(
                ' or '
            )}`
        );
    }

    const found: {
        jest: string[];
        jasmine: string[];
    } = { jest: [], jasmine: [] };

    const root = t.getDir('.');

    root.subfiles.forEach(f => {
        if(f.includes('node_modules')) {
            // skip node modules
            return;
        }

        if (f.includes('jest')) {
            found.jest.push(f);
            logger.debug(`Detected jest by way of ${f}`);
        } else if (f.includes('karma')) {
            found.jasmine.push(f);
            logger.debug(`Detected jasmine by way of: ${f}`);
        } else if (f.includes('package.json')) {
            const content = t.read(`./${f}`)?.toString();
            if (typeof content === 'string') {
                if (content.includes('"jasmine-core"')) {
                    found.jasmine.push(f);
                    logger.debug(`Detected jasmine by way of ${f}`);
                }
                if (content.includes('"jest"')) {
                    found.jest.push(f);
                    logger.debug(`Detected jest by way of ${f}`);
                }
            }
        }
    });

    if (found.jasmine.length > found.jest.length) {
        logger.debug(`Detected jasmine by way of ${found.jasmine.join()}`);
        return 'jasmine';
    } else if (found.jest.length > found.jasmine.length) {
        logger.debug(`Detected jest by way of ${found.jest.join()}`);
        return 'jest';
    } else {
        logger.debug(`Fallback to ${fallback}`);
        // jest == jasmine
        return fallback;
        // todo - read through the files again and get the .mockReturnValue  and '.and.' numbers and .returnValue counts to make sure which type of runner is being used
    }
}
