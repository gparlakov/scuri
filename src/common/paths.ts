import { normalize, basename, extname } from 'path';
import { getLogger } from './logger';
import { normalize as ngNormalize } from '@angular-devkit/core';

/**
 * Holds the folder, fileName and specFileName:
 *
 * @example for `./my/example/example.component.ts`
 * {
 *  folderPathRaw: './my/example/',
 *  folderPath: './my/example/,
 *  fileName: 'example.component.ts'
 *  specFileName: 'example.component.spec.ts'
 * }
 *
 * @example for `.\my\example\example.component.ts`
 * {
 *  folderPathRaw: '.\my\example\',,
 *  folderPath: './my/example/,
 *  testTarge
 *  fileName: 'example.component.ts',
 *  specFileName: 'example.component.spec.ts',
 * }
 *
 */
export type Paths = {
    /**
     * The file name path excluded
     * @example
     * --name = ./example/example.component.ts -> `example.component.spec.ts`
     */
    specFileName: string;

    /**
     * Import path with normalized slashes (/ and \ according to local OS)
     * @example
     * --name = ./example/example.component.ts -> `example.component` (notice the lack of extension)
     */
    fileName: string;

    /**
     * The folder containing the target under test, verbatim as the user specified.
     *
     * @example
     * --name = ./example/example.component.ts -> ./example/example.component
     *
     * @example
     * --name = .\example\example.component.ts -> .\example\example.component
     */
    folderPathRaw: string;
    /**
     * The folder containing the target under test, normalized to forward slash path.
     *
     * @example
     * --name = .\example\example.component.ts -> ./example/example.component
     *
     */
    folderPathNormal: string;
};

export function paths(name: string): Paths {
    const logger = getLogger(paths.name);
    logger.debug(`entering for ${name}`);
    // --name=./example/example.component.ts -> example.component.ts
    const fileName = basename(normalize(name));

    // normalize the / and \ according to local OS
    // --name = ./example/example.component.ts -> ./example/example.component and the ext name -> .ts
    // for import { ExampleComponent } from "./example/example.component"
    const fileNameSansExtension = fileName.slice(0, fileName.length - extname(fileName).length);

    // the new spec file name
    const specFileName = `${fileNameSansExtension}.spec.ts`;

    // the folder verbatim as the user specified
    const folderPathRaw = name.split(fileName)[0]; // split on the filename - so we get only an array of one item

    const folderPathNormal = ngNormalize(folderPathRaw);
    const r = {
        specFileName,
        fileName: fileNameSansExtension,
        folderPathRaw,
        folderPathNormal,
    };

    logger.debug(`returning ${JSON.stringify(r)}`)
    return r;
}
