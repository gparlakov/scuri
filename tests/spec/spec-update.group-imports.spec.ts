import { Tree } from '@angular-devkit/schematics';
import { SchematicTestRunner } from '@angular-devkit/schematics/testing';
import { filter } from 'rxjs/operators';
import { collectionPath } from './common';

describe('Calling update on existing specs without setup function', () => {
    let tree = Tree.empty();

    beforeEach(() => {
        tree = Tree.empty();
        tree.create(
            'c.ts',
            `import { MyDep } from '@angular/Common';
import { LibraryClass, LibraryClass1 } from '../shared/library';
import { LogService, BDep, CDep } from '@angular/core';

export class C  {
    constructor(
        private aDep: BDep,
        private bDep: BDep,
        private cDep: BDep,
        private dDep: CDep,
        private eDep: MyDep,
        private fDep: LibraryClass,
        private gDep: LibraryClass1,
        private logger: LogService
    ) {}
}`
        );

        tree.create(
            'c.spec.ts',
            `import { BDep, LogService } from '@angular/core';
import { other } from 'other';
import { some } from '..\\util';

describe('C', () => {
});`
        );
    });

    it('should pass successfully', async () => {
        // arrange
        const runner = new SchematicTestRunner('schematics', collectionPath);
        // act
        const errors = [];
        runner.logger.pipe(filter((v) => v.level === 'error')).subscribe((v) => errors.push(v));
        await runner.runSchematicAsync('spec', { name: './c.ts', update: true }, tree).toPromise();
        // assert
        expect(errors.length).toBe(0);
    });

    it('should import only missing deps and not duplicate deps (BDep is used multiple times), ', async () => {
        // arrange
        const runner = new SchematicTestRunner('schematics', collectionPath);
        // act
        const result = await runner
            .runSchematicAsync('spec', { name: './c.ts', update: true }, tree)
            .toPromise();
        // assert
        // @ts-ignore
        const contents = result.readContent('./c.spec.ts');
        expect(contents).toMatchInlineSnapshot(`
            "import { BDep, LogService, CDep } from '@angular/core';
            import { other } from 'other';
            import { some } from '..\\\\util';
            import { MyDep } from '@angular/Common';
            import { LibraryClass } from '../shared/library';
            import { LibraryClass1 } from '../shared/library';
            import { autoSpy } from 'autoSpy';

            describe('C', () => {
            });
            function setup() {
                const aDep = autoSpy(BDep);
                const bDep = autoSpy(BDep);
                const cDep = autoSpy(BDep);
                const dDep = autoSpy(CDep);
                const eDep = autoSpy(MyDep);
                const fDep = autoSpy(LibraryClass);
                const gDep = autoSpy(LibraryClass1);
                const logger = autoSpy(LogService);
                const builder = {
                    aDep,
                    bDep,
                    cDep,
                    dDep,
                    eDep,
                    fDep,
                    gDep,
                    logger,
                    default() {
                        return builder;
                    },
                    build() {
                        return new C(aDep, bDep, cDep, dDep, eDep, fDep, gDep, logger);
                    }
                }
                return builder;
            }"
        `);
    });

    it('should not import deps with no path i.e. from dom or other tslibs, Object, Event', async () => {
        // arrange
        const t = Tree.empty();
        t.create(
            'c.ts',
            `export class C  {
    constructor(
        private aDep: Event,
        private bDep: Object,
        private cDep: Window,
    ) {}
}`
        );

        t.create(
            'c.spec.ts',
            `describe('C', () => {
});`
        );
        const runner = new SchematicTestRunner('schematics', collectionPath);
        // act
        const result = await runner
            .runSchematicAsync('spec', { name: './c.ts', update: true }, t)
            .toPromise();
        // assert
        // @ts-ignore
        const contents = result.readContent('./c.spec.ts');
        expect(contents).toMatchInlineSnapshot(`
            "import { autoSpy } from 'autoSpy';
            describe('C', () => {
            });
            function setup() {
                const aDep = autoSpy(Event);
                let bDep: Object;
                const cDep = autoSpy(Window);
                const builder = {
                    aDep,
                    bDep,
                    cDep,
                    default() {
                        return builder;
                    },
                    build() {
                        return new C(aDep, bDep, cDep);
                    }
                }
                return builder;
            }"
        `);
    });
});
