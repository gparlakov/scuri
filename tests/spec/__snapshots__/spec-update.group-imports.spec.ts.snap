// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calling update on existing specs without setup function should import only missing deps and not duplicate deps (BDep is used multiple times),  1`] = `
"import { BDep, LogService, CDep } from '@angular/core';
import { other } from 'other';
import { some } from '../util';
import { MyDep } from '@angular/Common';
import { LibraryClass } from '../shared/library';
import { LibraryClass1 } from '../shared/library';
import { autoSpy } from 'autoSpy';

describe('C', () => {
});

function setup() {
    const aDep = autoSpy(BDep);
    const bDep = autoSpy(BDep);
    const cDep = autoSpy(BDep);
    const dDep = autoSpy(CDep);
    const eDep = autoSpy(MyDep);
    const fDep = autoSpy(LibraryClass);
    const gDep = autoSpy(LibraryClass1);
    const logger = autoSpy(LogService);
    const builder = {
        aDep,
        bDep,
        cDep,
        dDep,
        eDep,
        fDep,
        gDep,
        logger,
        default() {
            return builder;
        },
        build() {
            return new C(aDep, bDep, cDep, dDep, eDep, fDep, gDep, logger);
        },
        
    }
    return builder;
}"
`;

exports[`Calling update on existing specs without setup function should not import deps with no path i.e. from dom or other tslibs, Object, Event 1`] = `
"import { autoSpy } from 'autoSpy';
describe('C', () => {
});

function setup() {
    const aDep = autoSpy(Event);
    let bDep: Object;
    const cDep = autoSpy(Window);
    const builder = {
        aDep,
        bDep,
        cDep,
        default() {
            return builder;
        },
        build() {
            return new C(aDep, bDep, cDep);
        },
        
    }
    return builder;
}"
`;
